eclipse-theia/theia
# Refactor auto save mechanism

#### What it does

Closes https://github.com/eclipse-theia/theia/issues/12863
Closes https://github.com/eclipse-theia/theia/issues/12844
Closes https://github.com/eclipse-theia/theia/issues/8722

Enhances the `SaveResourceService` to perform all the auto save preference work. Previously, all editor classes had to do their own handling of the `afterDelay` auto save functionality. This has been centralized into one service. Therefore, custom editors and notebook editors now behave exactly the same as monaco text editors in regards to auto saves.

This also fixes a long standing issue, where `onFocusChange` and `onWindowChange` have been handled as if `afterDelay` was selected as the auto save mode. These two values are now properly handled.

Note that `onWindowChange` doesn't work well with custom editors. This is due to the usage of `iframe` and the inability to look into an iframe. It still works, but is a bit too eager with them.

#### How to test

Basically, confirm that the auto save functionality behaves as expected with all different values of the enum preference. This should be tested for all 3 kinds of editors (text, notebook, custom editor).

Also test that closing unsaved editors prompts (or automatically saves) the user whether or not to save those files.

#### Follow-ups

We might want to find a way to fix `onWindowChange` for custom editors as well based on mouse position event information.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
